cmd/kubeadm/app/util/config/initconfiguration.go

```diff
@@ -112,6 +114,10 @@ func SetNodeRegistrationDynamicDefaults(cfg *kubeadmapi.NodeRegistrationOptions,
 		if err != nil {
 			return err
 		}
+		// Using "/run/containerd/containerd.sock" as endpoint is deprecated, use full url format "unix:///run/containerd/containerd.sock".
+		if filepath.IsAbs(cfg.CRISocket) && goruntime.GOOS != "windows" {
+			cfg.CRISocket = "unix://" + cfg.CRISocket
```

Comment:
Thanks,
This is ok for linux, but i don't like that we are skipping windows here.
The abs function does not work for windows if the path does not start with a drive letter.
so it will be already false even without the runtime check.

We need a util function in separete _windows _linux files that does a different thing.

The other question here is - should we just start showing warnings instead of prepending the socket paths?

Reply:
Agree that we should handle the case for windows. 

>       --container-runtime-endpoint string                        [Experimental] The endpoint of remote runtime service. Currently unix socket endpoint is supported on Linux, while npipe and tcp endpoints are supported on windows.  Examples:'unix:///var/run/dockershim.sock', 'npipe:////./pipe/dockershim' (default "unix:///var/run/dockershim.sock")


> The other question here is - should we just start showing warnings instead of prepending the socket paths?

/sig node
The warning comes from kubelet #46089.
Defaulting to something kubelet prefers would be better.

Or we should change `function detectCRISocketImpl` to match it. 
https://github.com/kubernetes/kubernetes/blob/cea1d4e20b4a7886d8ff65f34c6d4f95efcb4742/cmd/kubeadm/app/util/runtime/runtime.go#L195-L229

Reply:
to detect a socket, on windows we have isExistingSocket that dials the named pipe.
the function for windows does a stat on file on disk.

we could modify all default sockets on linux to be unix:///somepath
https://github.com/kubernetes/kubernetes/blob/2b7c61b1bb1889770f8d8cbdf0119617c5228f6e/cmd/kubeadm/app/constants/constants_unix.go#L23
https://github.com/kubernetes/kubernetes/blob/cea1d4e20b4a7886d8ff65f34c6d4f95efcb4742/cmd/kubeadm/app/util/runtime/runtime_unix.go#L26-L27
https://github.com/kubernetes/kubernetes/blob/cea1d4e20b4a7886d8ff65f34c6d4f95efcb4742/cmd/kubeadm/app/util/runtime/runtime.go#L200

then in the linux version of isExistingSocket we could split the unix:// and sometpath
and do `_, err := net.Dial("unix", "somepath")`

for `err == nil` we found a socket.

this would allow us to always default with the unix:// in the constants on Linux too.

and this will remove the warning from the kubelet.

changing API defaults on dynamic defaulting is fine, but for the static defaulting (DefaultDockerCRISocket) this is not great:
https://github.com/kubernetes/kubernetes/blob/645c40fcf6f1fca133a00c8186674bcbcecc4b8e/cmd/kubeadm/app/util/config/initconfiguration.go#L183-L190

but i think i'm OK with the change given it causes warnings.

Reply:
> then in the linux version of isExistingSocket we could split the unix:// and sometpath
> and do _, err := net.Dial("unix", "somepath")
> for err == nil we found a socket.
> ^ has to be tested.

[1] this works:
```
package main

import (
	"fmt"
	"net"
	"net/url"
)

func main() {
	fmt.Println(isExistingSocket("unix:///var/run/dockershim.sock"))
	fmt.Println(isExistingSocket("unix:///foo/bar"))
}

// isExistingSocket checks if path exists and is domain socket
func isExistingSocket(path string) bool {
	u, err := url.Parse(path)
	if err != nil {
		return false
	}
	c, err := net.Dial(u.Scheme, u.Path)
	if err != nil {
		return false
	}
	defer c.Close()
	return true
}
```

but it requires root to dial an actual socket. should be fine as long as kubeadm commands not requiring root don't run the socket dynamic defaulting.

[2] if that is a concert we can preserve the old method, but extend it to parse full URIs:

```
// isExistingSocket checks if path exists and is domain socket
func isExistingSocket(path string) bool {
	u, err := url.Parse(path)
	if err != nil {
		return false
	}
	fileInfo, err := os.Stat(u.Path)
	if err != nil {
		return false
	}

	return fileInfo.Mode()&os.ModeSocket != 0
}
```

i prefer option [1].

Reply:
```
package main

import (
	"fmt"
	"net"
	"net/url"
)

func main() {
	fmt.Println(isExistingSocket("unix:///var/run/dockershim.sock"))
	fmt.Println(isExistingSocket("unix:///foo/bar"))
        fmt.Println(isExistingSocket("/var/run/containerd/containerd.sock"))
        fmt.Println(isExistingSocket("/var/run/docker.sock"))
}

// isExistingSocket checks if path exists and is domain socket
func isExistingSocket(path string) bool {
	u, err := url.Parse(path)
	if err != nil {
		return false
	}
	if u.Scheme == "" {
		// add a warning log here, if user config it to be a path, not a url.
		u.Scheme = "unix"
	}
	c, err := net.Dial(u.Scheme, u.Path)
	if err != nil {
		return false
	}
	defer c.Close()
	return true
}

```